python/dgl/id_map.py

```diff
@@ -0,0 +1,68 @@
+"""Module for mapping between node/edge Ids and node/edge types."""
+import numpy as np
+
+from ._ffi.function import _init_api
+from . import backend as F
+from . import utils
+
+class IdMap:
+    '''This maps node/edge Ids in the homogeneous form to per-type form.
+
+    It stores the Id ranges for each partition and for each type. The Ids are in
+    the homogeneous form (i.e., all nodes of different types have unique Ids). For each type,
+    all Ids in a partition fall in a contiguous range. The Id ranges are stored in
+    a matrix, where each row has two elements to represent the beginning and the end of
+    the range.
```

Comment:
Even with the help of your doc, it is very hard to decipher this description. I could only make the following guess:
1. The input `id_ranges` is an object of type `dict[str, Tensor]`. It stores the same information as the `node_map` argument to `GraphPartitionBook`.
2. The input ranges are global ranges, meaning that it refers to ID ranges w.r.t. the entire graph.
3. The goal of this `IdMap` class is to parse a global node ID into its type ID and also the offset within the type (or called per-type ID). The function should support vectorization, i.e., process multiple ID conversions in parallel.

If my understanding is correct, we could formulate the problem as the following. Given a list of sorted values `A` and a list of arbitrary values `B`, return an index `i` for each element in `B` such that `A[i] <= the_element < A[i + 1]`. If we replace `A` with the cumsum array of the number of nodes of each type globally and `B` with the IDs to convert, the result is the type ID of each input. This is exactly [`numpy.searchsorted`](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html). PyTorch added this op in 1.6.0. Getting the offset (per-type ID) of each element is also achievable using array operations:
```python
A = ...  # zero-based cumsum(num_nodes_per_type)
B = ...  # IDs to convert
I = searchsorted(A, B)  # type ID of each element in B
per_type_id = B - A[I]
```

Reply:
I think it's more complex than that because we need to deal with partitions as well. Here is an example:
```
+---------+-----+----------
  range   | type | partition
(0, 100)  |   0    |    0
(100,200) |   1    |    0
(200,300) |   0    |    1
(300,400) |   1    |    1
```
With binary search, we can get its type, but getting per-type ID is a little more difficult.
I guess it's still doable with pure tensor operations. let me try